### PR TITLE
feat: support custom global analytics key

### PR DESCRIPTION
**Pull request by oscb**
- [x] I've included a changeset (psst. run `yarn changeset`. Read about changesets [here](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md)).

### CODE REVIEW

**Comment by silesky** [comment]
Hey @oscb -- slightly confused by this PR -- I thought we want to allow users to use window.foo.track() rather than window.analytics.track()? So the config option would be like 'analyticsGlobalKey'

**Comment by oscb** [comment]
> I thought the goal was to allow users to use window.foo.track() rather than window.analytics.track()? So the config option would be like 'analyticsGlobalKey', and we save that key in memory -- using it whenever we want to get the global analytics, like how getCDNUrl() works. The fact that this would also be the same reference as the buffer is just an implementation detail.

Yes, it's sort of the idea. I think the name of the config option can be changed to `analyticsGlobalKey`

Per my discussion with Chris we're not yet changing the script part of it as it is going to be a challenge given that the script itself uses `window.analytics` but devs can modify the script and use their custom key and send it through the config (needs docs).

File: .changeset/mean-apricots-hammer.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+'@segment/analytics-next': minor
+'@internal/consent-tools-integration-tests': minor
+---
+
+Adds `bufferKey` option for setting custom global window buffers
```
Comment (silesky) [inline .changeset/mean-apricots-hammer.md:6]: globalAnalyticsKey

File: packages/browser/src/browser/__tests__/integration.test.ts [hunk 2]
```diff
@@ -212,6 +213,57 @@ describe('Initialization', () => {
     })
   })
 
+  describe('bufferKey', () => {
```
Comment (silesky) [inline packages/browser/src/browser/__tests__/integration.test.ts:216]: globalAnalyticsKey

File: packages/core/src/task/__tests__/task-group.test.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { createTaskGroup } from '../task-group'
```
Comment (silesky) [inline packages/core/src/task/__tests__/task-group.test.ts:1]: thanks!

File: packages/browser/src/core/analytics/index.ts [hunk 4]
```diff
@@ -117,6 +117,11 @@ export interface InitOptions {
    * Array of high entropy Client Hints to request. These may be rejected by the user agent - only required hints should be requested.
    */
   highEntropyValuesClientHints?: HighEntropyHint[]
+  /**
+   * Key for the global window property storing the buffered calls
```
Comment (silesky) [inline packages/browser/src/core/analytics/index.ts:121]:
```suggestion
   * When using the snippet, this is the key that points to the global analytics instance (e.g. window.analytics).
```

File: packages/browser/src/browser/index.ts [hunk 5]
```diff
@@ -162,11 +164,12 @@ async function flushFinalBuffer(
   analytics: Analytics,
   buffer: PreInitMethodCallBuffer
 ): Promise<void> {
+  const calls = getGlobalAnalytics()
```
Comment (silesky) [inline packages/browser/src/browser/index.ts:167]: Can we get rid of this now? restore the old function signature of "popSnippetWindowBuffer"?

File: packages/browser/src/browser/standalone-analytics.ts [hunk 6]
```diff
@@ -59,15 +45,15 @@ function getWriteKey(): string | undefined {
...
       'Failed to load Write Key. Make sure to use the latest version of the Segment snippet, which can be found in your source settings.'
     )
     return
   }
 
-  window.analytics = (await AnalyticsBrowser.standalone(
+  ;(window as any).analytics = (await AnalyticsBrowser.standalone(
     writeKey,
     options
   )) as AnalyticsSnippet
```
Comment (silesky) [inline packages/browser/src/browser/standalone-analytics.ts:59]: Should this be something like this:
```ts
AnalyticsBrowser.standalone(
    writeKey,
    options
  )).then(analytics => {
       window[options?.globalAnalyticsKey || 'analytics'] = analytics
  })
```

File: packages/browser/src/core/analytics/index.ts [hunk 7]
```diff
@@ -489,7 +494,7 @@ export class Analytics
 
   noConflict(): Analytics {
     console.warn(deprecationWarning)
-    window.analytics = _analytics ?? this
+    ;(window as any).analytics = _analytics ?? this
```
Comment (silesky) [inline packages/browser/src/core/analytics/index.ts:497]: Should this be:
``` 
setGlobalAnalytics(_analytics ?? this)
```

File: packages/browser/src/browser/utils.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * Gets the global analytics instance/buffer
+ * @param key name of the window property where the buffer is stored (default: analytics)
+ * @returns AnalyticsSnippet
+ */
+export function getGlobalAnalytics(): AnalyticsSnippet | undefined {
+  return (window as any)[_globalAnalyticsKey]
+}
+
```
Comment (silesky) [inline packages/browser/src/browser/utils.ts:16]: Can we add this?
```ts
export function setGlobalAnalytics(analytics: AnalyticsSnippet): void {
   (window as any)[_globalAnalyticsKey] = analytics
}
```

**Review by silesky — CHANGES_REQUESTED** [review]

File: packages/browser/src/browser/index.ts (hunk 5 above)
Comment (oscb) [reply packages/browser/src/browser/index.ts:167]: Yea I guess we don't ned this anymore since we store it globally

**Review by silesky — APPROVED** [review]

**Review by silesky — CHANGES_REQUESTED** [review]
Small late breaking feedback just noticed!

File: packages/browser/src/browser/utils.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import { AnalyticsSnippet } from './standalone-interface'
```
Comment (silesky) [inline packages/browser/src/browser/utils.ts:1]: Since file has import side effects -- I think we should name it 'global-analytics-helpers.ts ' or something like that, so nobody puts anything else in it. It could also go in /lib

File: packages/browser/src/browser/utils.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Stores the global window analytics key
+ */
+let _globalAnalyticsKey = 'analytics'
+
+/**
+ * Gets the global analytics instance/buffer
+ * @param key name of the window property where the buffer is stored (default: analytics)
```
Comment (silesky) [inline packages/browser/src/browser/utils.ts:10]: Can we change the comment to say "global analytics/buffer"

File: packages/browser/src/browser/utils.ts (hunk 9 above)
Comment (oscb) [reply packages/browser/src/browser/utils.ts:1]: Good point. It stopped being only an `utils` file

**Review by silesky — APPROVED** [review]

File: .changeset/mean-apricots-hammer.md [hunk 11]
```diff
@@ -0,0 +1,6 @@
+---
+'@segment/analytics-next': minor
+'@internal/consent-tools-integration-tests': minor
```
Comment (silesky) [inline .changeset/mean-apricots-hammer.md:3]: this looks like a mistake

File: packages/browser/src/browser/standalone-analytics.ts [hunk 12]
```diff
@@ -1,30 +1,19 @@
-import { Analytics, InitOptions } from '../core/analytics'
 import { AnalyticsBrowser } from '.'
```
Comment (silesky) [inline packages/browser/src/browser/standalone-analytics.ts:1]: Will changing this file name change the structure of the UMD build folder? If so, I would do this in a separate PR where we could evaluate the implications, as this is an actual entry point to the app and I would consider it to be a public API.

**Review by silesky — CHANGES_REQUESTED** [review]
Some other stuff ;-)

File: packages/browser/src/tester/ajs-tester.ts [hunk 13]
```diff
@@ -12,11 +12,13 @@ function makeStub(page: playwright.Page) {
       ...args: Parameters<Analytics['register']>
     ): Promise<SerializedContext> {
       return await page.evaluate((innerArgs) => {
-        // @ts-ignore
-        return window.analytics
-          .register(...innerArgs)
-          .then((ctx) => ctx.toJSON())
-        // @ts-ignore
+        return (
```
Comment (silesky) [inline packages/browser/src/tester/ajs-tester.ts:15]: It's actually fine to do declare global  in test code, since it doesn't get included in the actual published library.

File: examples/with-next-js/pages/partytown/index.tsx [hunk 14]
```diff
@@ -44,7 +44,7 @@ const WebWorker: React.FC = () => {
           }}
           value="Track!"
```
Comment (silesky) [inline examples/with-next-js/pages/partytown/index.tsx:45]: It's actually fine to do:
```
declare global {
  interface Window {
    analytics: AnalyticsSnippet
  }
}
```
... since this is not a published repo

Comment (oscb) [reply examples/with-next-js/pages/partytown/index.tsx:45]: I'll leave it as it is for now. But yeah, doesn't matter.

**Review by silesky — APPROVED** [review]
